File: frontends/solid/src/stores/plaza.ts
// ============================================================
// 广场状态：sort / filter / 关键词 + 分页 + 列表缓存
// ============================================================

import { createStore } from "solid-js/store";
import { api } from "@/api/client";
import type {
  CapsuleListItem,
  Pagination,
  PlazaFilter,
  PlazaSort,
} from "@/types";

interface PlazaStore {
  sort: PlazaSort;
  filter: PlazaFilter;
  q: string;
  page: number;
  pageSize: number;

  items: CapsuleListItem[];
  pagination: Pagination | null;
  loading: boolean;
  error: string | null;
}

const [plaza, setPlaza] = createStore<PlazaStore>({
  sort: "new",
  filter: "all",
  q: "",
  page: 1,
  pageSize: 15,

  items: [],
  pagination: null,
  loading: false,
  error: null,
});

export { plaza };

// 请求序列号：fetch 内仅在自身是「最新发起」时才回写结果，避免乱序覆盖。
// 用闭包变量而非 store state，避免触发额外订阅。
let fetchSeq = 0;

export async function fetchPlaza() {
  const myId = ++fetchSeq;
  setPlaza({ loading: true, error: null });
  try {
    const data = await api.plaza({
      sort: plaza.sort,
      filter: plaza.filter,
      q: plaza.q.trim() || undefined,
      page: plaza.page,
      pageSize: plaza.pageSize,
    });
    // 落后的请求直接丢弃，避免覆盖更新的结果
    if (myId !== fetchSeq) return;
    setPlaza({ items: data.items, pagination: data.pagination, loading: false });
  } catch (e) {
    if (myId !== fetchSeq) return;
    setPlaza({
      loading: false,
      error: e instanceof Error ? e.message : "加载失败",
    });
  }
}

export function setSort(s: PlazaSort) {
  setPlaza({ sort: s, page: 1 });
  void fetchPlaza();
}
export function setFilter(f: PlazaFilter) {
  setPlaza({ filter: f, page: 1 });
  void fetchPlaza();
}
export function setQ(q: string) {
  setPlaza({ q, page: 1 });
  void fetchPlaza();
}
export function setPage(p: number) {
  setPlaza({ page: p });
  void fetchPlaza();
}

export function patchPlazaFavorited(
  capsuleId: string,
  favorited: boolean,
  count: number,
) {
  setPlaza("items", (items) =>
    items.map((it) =>
      it.id === capsuleId
        ? { ...it, favoritedByMe: favorited, favoriteCount: count }
        : it,
    ),
  );
}

File: desktop/qt-pyside/app/stores/plaza.py
"""广场 store：sort/filter/q + 分页 + 列表（= React stores/plaza.ts）。fetch 用序列号防乱序。"""
from __future__ import annotations

from PySide6.QtCore import Property, QObject, Signal, Slot

from api_client import ApiClient
from worker import run_async


class PlazaStore(QObject):
    changed = Signal()

    def __init__(self, api: ApiClient):
        super().__init__()
        self._api = api
        self._sort = "new"
        self._filter = "all"
        self._q = ""
        self._page = 1
        self._page_size = 15
        self._items: list = []
        self._total = 0
        self._total_pages = 0
        self._loading = False
        self._seq = 0

    @Property("QVariant", notify=changed)
    def items(self):
        return self._items

    @Property(str, notify=changed)
    def sort(self):
        return self._sort

    @Property(str, notify=changed)
    def filter(self):
        return self._filter

    @Property(int, notify=changed)
    def page(self):
        return self._page

    @Property(int, notify=changed)
    def total(self):
        return self._total

    @Property(int, notify=changed)
    def totalPages(self):
        return self._total_pages

    @Property(bool, notify=changed)
    def loading(self):
        return self._loading

    @Slot(str)
    def setSort(self, s: str):
        self._sort = s
        self._page = 1
        self.changed.emit()
        self.fetch()

    @Slot(str)
    def setFilter(self, f: str):
        self._filter = f
        self._page = 1
        self.changed.emit()
        self.fetch()

    @Slot(str)
    def setQ(self, q: str):
        self._q = q
        self._page = 1
        self.changed.emit()
        self.fetch()

    @Slot(int)
    def setPage(self, p: int):
        self._page = p
        self.changed.emit()
        self.fetch()

    @Slot()
    def fetch(self):
        self._seq += 1
        my = self._seq
        self._loading = True
        self.changed.emit()

        def ok(data):
            if my != self._seq:
                return
            self._items = data.get("items", [])
            pg = data.get("pagination", {})
            self._total = pg.get("total", 0)
            self._total_pages = pg.get("totalPages", 0)
            self._loading = False
            self.changed.emit()

        def err(_):
            if my != self._seq:
                return
            self._loading = False
            self.changed.emit()

        run_async(
            self._api.plaza, ok, err,
            self._sort, self._filter, self._q.strip() or None, self._page, self._page_size,
        )

    def patchFavorited(self, capsule_id: str, favorited: bool, count: int):
        changed = False
        new_items = []
        for it in self._items:
            if it.get("id") == capsule_id:
                it = {**it, "favoritedByMe": favorited, "favoriteCount": count}
                changed = True
            new_items.append(it)
        if changed:
            self._items = new_items
            self.changed.emit()
